packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,605 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import 'dart:ui' show lerpDouble;
+
+/// A class that abstracts over a Material slice and a Material gap.
+///
+/// All [MergeableMaterialItem] objects need a [LocalKey].
+abstract class MergeableMaterialItem {
+  MergeableMaterialItem(this.key) {
+    assert(key != null);
+  }
+
+  final LocalKey key;
+}
+
+/// A class that can be used as a child to [MergeableMaterial]. It is a slice
+/// of [Material] that animates merging with other slices.
+///
+/// All [MaterialSlice] objects need a [LocalKey].
+class MaterialSlice extends MergeableMaterialItem {
+  /// Creates a slice of [Material] that's mergeable within a
+  /// [MergeableMaterial].
+  MaterialSlice({
+    LocalKey key,
+    this.child
+  }) : super(key);
+
+  /// The contents of this slice.
+  final Widget child;
+
+  @override
+  String toString() {
+    return 'MergeableSlice(key: $key, child: $child)';
+  }
+}
+
+/// A class that represents a gap within [MergeableMaterial].
+///
+/// All [MaterialGap] objects need a [LocalKey].
+class MaterialGap extends MergeableMaterialItem {
+  /// Creates a Material gap with a given size.
+  MaterialGap({
+    LocalKey key,
+    this.size: 16.0
+  }) : super(key);
+
+  /// The main axis extent of this gap. For example, if the [MergableMaterial]
+  /// is vertical, then this is the height of the gap.
+  final double size;
+
+  @override
+  String toString() {
+    return 'MaterialGap(key: $key, child: $size)';
+  }
+}
+
+/// A widget that animates the merging and separating of [Material] slices
```

Comment:
This is an important part of the documentation since it explains what all the parts are for and how they fit together. Here's a suggestion:

Displays a list of [MergeableMaterialItem] children. The list contains [MaterialSlice] items whose boundaries are either "merged" with adjacent items or separated by a [MaterialGap]. The [children] are distributed along the given [mainAxis] in the same way as the children of a [Block]. When the list of children changes, gaps are automatically animated open or closed as needed.

To enable this widget to correlate its list of children with the previous one, each child must specify a key.

When a new gap is added to the list of children the adjacent items are animated apart. Similarly when a gap is removed the adjacent items are brought back together.

When a new slice is added or removed the app is responsible for animating the transition.